pkg/controller/pxc/upgrade.go

```diff
@@ -191,41 +192,80 @@ func (r *ReconcilePerconaXtraDBCluster) smartUpdate(sfs api.StatefulApp, cr *api
 			primaryPod = pod
 		} else {
 			log.Info(fmt.Sprintf("apply changes to secondary pod %s", pod.Name))
-			if err := r.applyNWait(cr, sfs.StatefulSet().Status.UpdateRevision, &pod, waitLimit); err != nil {
+			if err := r.applyNWait(cr, sfs.StatefulSet(), &pod, waitLimit); err != nil {
 				return fmt.Errorf("failed to apply changes: %v", err)
 			}
 		}
 	}
 
 	log.Info(fmt.Sprintf("apply changes to primary pod %s", primaryPod.Name))
-	if err := r.applyNWait(cr, sfs.StatefulSet().Status.UpdateRevision, &primaryPod, waitLimit); err != nil {
+	if err := r.applyNWait(cr, sfs.StatefulSet(), &primaryPod, waitLimit); err != nil {
 		return fmt.Errorf("failed to apply changes: %v", err)
 	}
 
 	return nil
 }
 
-func (r *ReconcilePerconaXtraDBCluster) applyNWait(cr *api.PerconaXtraDBCluster, updateRevision string, pod *corev1.Pod, waitLimit int) error {
-	if pod.ObjectMeta.Labels["controller-revision-hash"] == updateRevision {
+func (r *ReconcilePerconaXtraDBCluster) applyNWait(cr *api.PerconaXtraDBCluster, sfs *appsv1.StatefulSet, pod *corev1.Pod, waitLimit int) error {
+	if pod.ObjectMeta.Labels["controller-revision-hash"] == sfs.Status.UpdateRevision {
 		log.Info(fmt.Sprintf("pod %s is already updated", pod.Name))
 	} else {
 		if err := r.client.Delete(context.TODO(), pod); err != nil {
 			return fmt.Errorf("failed to delete pod: %v", err)
 		}
 	}
 
-	if err := r.waitPodRestart(updateRevision, pod, waitLimit); err != nil {
+	if err := r.waitPodRestart(sfs.Status.UpdateRevision, pod, waitLimit); err != nil {
 		return fmt.Errorf("failed to wait pod: %v", err)
 	}
 
+	if err := r.waitUntilOnline(cr, pod, waitLimit); err != nil {
+		return fmt.Errorf("failed to wait pxc status: %v", err)
+	}
+
 	if err := r.waitPXCSynced(cr, pod.Status.PodIP, waitLimit); err != nil {
 		return fmt.Errorf("failed to wait pxc sync: %v", err)
 	}
 
 	return nil
 }
 
-func (r *ReconcilePerconaXtraDBCluster) getPrimaryPod(cr *api.PerconaXtraDBCluster) (string, error) {
+func (r *ReconcilePerconaXtraDBCluster) waitUntilOnline(cr *api.PerconaXtraDBCluster, pod *corev1.Pod, waitLimit int) error {
+	database, err := r.proxyDB(cr)
```

Comment:
can we do this magic only if ProxySQL is enabled?

Reply:
We have check for that
https://github.com/percona/percona-xtradb-cluster-operator/blob/master/pkg/apis/pxc/v1/pxc_types.go#L473

Reply:
btw, there is could be panic instead of error (nil pointer dereference of cr.Spec.ProxySQL). I asked to fix it in https://github.com/percona/percona-xtradb-cluster-operator/pull/443#discussion_r447715288. There it is almost 100% case on the legit path.